### DISCUSSION TITLE
"this" as a Built in Generic

### DISCUSSION THREAD

**Discussion by SteffenBlake** [dotnet/csharplang#6960] [Language Ideas]
There exists an awkward "hole" in designing Interfaces/Classes at this time for C# that can be filled by one of two very awkward workarounds.

I would propose a simple `this` keyword Type be introduced that acts as a "baked in" generic any class/interface may use, that simply is the constraint of "Is an instance of this type"

Let me provide an example, consider the following

```
abstract class MyBaseClass {

    public MyBaseClass Create() {
            ...
    }
}
class MyDerivedClass : MyBaseClass { ... }

// Main
var foo = new MyDerivedClass(); // foo's type is a MyDerivedClass 
var bar = foo.Create(); // Returns a MyBaseClass 
```

But what if you want a method on the class which returns one of `whatever it is`, but it's predefined by the base abstraction. A common example is wanting to make Fluent style interfaces that do something like:

```
var result = foo.Bar().Baz().Etc();
```
Where all the above methods return the class itself so you can chain methods together, but what if the expectation of your framework is the consumer makes their own implementation of a `foo`, such that your class is just an abstract class with a bunch of helper methods baked into it, and the consumer just needs to implement some form of core mechanism and all the "chaining" methods work.

Well, you can't do it easily but there exist 2 workarounds that both have annoying downsides.

# Option 1: Self Referential Generic

This is the first approach, and it sort of works but it can quickly become difficult to maintain if the same class also has a multitude of other generics, as each class that then refers to your class also has to refer to every *other* generic it has, which can quickly balloon to be a big mess.

It usually ends up something like this:

```
abstract class BaseFoo<TFoo>
     where TFoo : BaseFoo<TFoo>
{
      public TFoo Bar() { ... }
      public TFoo Baz() { ... }
      public TFoo Etc() { ... }
}

class Foo : BaseFoo<Foo> { ... }
```
The above methods can now do something like `return this` and it will work as expected, but as mentioned prior, if your `BaseFoo` has other generics added to it you now need to handle that in every single class that refers to it, and that can be a pain

# Option 2: Static Extension Methods

This is a simpler workaround but it effectively just results in a whole lot of extra cruft *just* so you can make your methods chainable.

Step one is to instead make your class have *non-chaining* versions of the methods like so:

```
abstract class BaseFoo
{
      public void Bar() { ... }
      public void Baz() { ... }
      public void Etc() { ... }
}
```
And then you simply make static extensions like so:

```
static class BaseFooExtensions
{
      public TFoo BarChainable<TFoo>(this TFoo foo) 
           where TFoo : BaseFoo
       { 
             foo.Bar()
             return this;
       }

      public TFoo BazChainable<TFoo>(this TFoo foo) 
           where TFoo : BaseFoo
       { 
             foo.Baz()
             return this;
       }

      public TFoo EtcChainable<TFoo>(this TFoo foo) 
           where TFoo : BaseFoo
       { 
             foo.Etc()
             return this;
       }
}
```

Which once again, it *works* but thats a lot of extra boilerplate for simple functionality.

# Proposed Solution: "this" generic type

I would propose the `this` keyword can be used as a Type for class/interface signatures, and it would work as simple as so, for Properties, Method args, and return types:

```
abstract class BaseFoo
{
      public this? Component { get; }

      public BaseFoo(this? component) 
      {
           Component = component;
      }

      public this Bar() { ... }
      public this Baz() { ... }
      public this Etc() { ... }
}
class Foo : BaseFoo { ... }

Foo myFoo = new Foo();

Foo anotherFoo = myFoo.Bar().Baz().Etc();
```

Much cleaner and removes the necessity for inversion of control via static extensions, the code is way easier to follow, and we dont have to reach for weird self referential generics.

**Comment by bernd5** [comment]
See: https://github.com/dotnet/csharplang/blob/711b9406f07d1fc43d2825df4c0284c9614b7821/proposals/self-constraint.md

**Reply by svick** [reply]
The associated issue is at https://github.com/dotnet/csharplang/issues/5413. (Especially note that this particular proposal was rejected.)
